diesel/src/query_source/joins.rs

```diff
@@ -21,22 +31,30 @@ impl<Left, Right, Kind> Join<Left, Right, Kind> {
             kind: kind,
         }
     }
+
+    #[doc(hidden)]
+    pub fn on<On>(self, on: On) -> JoinOn<Self, On> {
+        JoinOn {
+            join: self,
+            on: on,
```

Comment:
I've been shying away from adding those constraints earlier than needed unless there's a concrete case where it improves errors. I'm not sure that it would do so here.